EIPS/eip-3336.md

```diff
@@ -0,0 +1,71 @@
+---
+eip: 3336
+title: Paged memory allocation for the EVM
+author: Nick Johnson (@arachnid)
+discussions-to: https://ethereum-magicians.org/t/eips-3336-and-3337-improving-the-evms-memory-model/5482
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-03-06
+---
+
+## Simple Summary
+Changes the memory model for the EVM to use pagination, which permits more flexible use of memory by compilers without unduly complicating implementations.
+
+## Abstract
+Presently, the EVM charges for memory as a linear array starting at address 0 and extending to the highest address that has been read from or written to. This suffices for simple uses, but means that compilers have to generate programs that use memory compactly, which leads to wasted gas with reallocation of memory elements, and makes some memory models such as separate heap and stack areas impractical. This EIP proposes changing to a page-based billing model, which adds minimal complexity to implementations, while providing for much more versatility in EVM programs.
```

Comment:
The abstract should be a terse human readable version of the specification.  Currently, this discusses history which is more appropriate for the Motivation section.

(normally I would try to provide an alternative abstract, but I don't think I grok the change well enough to do so... sorry!)

Reply:
```suggestion
This EIP proposes changing EVM memory to a page-based allocation and gas billing model by dividing the memory address space into 1 kilobyte (32 word) pages, and charging gas on the basis of the number of pages allocated, rather than the address of the highest word referenced in memory. This facilitates more flexible use of memory than is currently possible with a linear model.
```

How about this?

Reply:
EIPs don't propose, they assert.  Also recommend removing the Rationale at the end and keeping focused on the specification.
```suggestion
Makes EVM memory a page-based allocation system and gas billing model by dividing the memory address space into 1 kilobyte (32 word) pages, and charging gas on the basis of the number of pages allocated, rather than the address of the highest word referenced in memory.
```